File: src/verify_refs.py
"""Verify every reference in a BibTeX file against Crossref, and every citation against the file.

For each entry with a DOI, query https://api.crossref.org/works/<doi> and compare:
  title (normalized similarity >= 0.90), first author family name, year, journal/container.
Mismatches fail loudly (nonzero exit). Entries without a DOI are listed for manual checking.
If a .tex file is given, every \\cite key must exist in the .bib and every .bib entry must be cited.

Usage: python src/verify_refs.py paper/references.bib [paper/itoju.tex]
"""
import difflib
import json
import re
import sys
import time
import unicodedata
import urllib.parse
import urllib.error
import urllib.request
from pathlib import Path

MAILTO = "itoju-reference-check"


def norm(s):
    s = unicodedata.normalize("NFKD", s or "")
    s = "".join(c for c in s if not unicodedata.combining(c))
    s = re.sub(r"\\[a-zA-Z]+|[{}\\$]", "", s)
    return re.sub(r"[^a-z0-9 ]+", " ", s.lower()).split()


def similar(a, b):
    return difflib.SequenceMatcher(None, " ".join(norm(a)), " ".join(norm(b))).ratio()


def parse_bib(text):
    entries = {}
    for m in re.finditer(r"@(\w+)\s*\{\s*([^,]+),(.*?)\n\}", text, flags=re.S):
        kind, key, body = m.group(1).lower(), m.group(2).strip(), m.group(3)
        fields = {}
        for fm in re.finditer(r"(\w+)\s*=\s*(\{(?:[^{}]|\{(?:[^{}]|\{[^{}]*\})*\})*\}|\"[^\"]*\"|\d+)", body, flags=re.S):
            v = fm.group(2).strip()
            if v[0] in "{\"":
                v = v[1:-1]
            fields[fm.group(1).lower()] = re.sub(r"\s+", " ", v).strip()
        entries[key] = {"type": kind, **fields}
    return entries


def crossref(doi):
    """Crossref record, or a DataCite record mapped to the same fields (repository DOIs such as
    Stanford Digital Repository preprints are registered with DataCite, not Crossref)."""
    url = "https://api.crossref.org/works/" + urllib.parse.quote(doi)
    req = urllib.request.Request(url, headers={"User-Agent": f"verify_refs/1.0 ({MAILTO})"})
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            return json.load(r)["message"]
    except urllib.error.HTTPError as ex:
        if ex.code != 404:
            raise
    with urllib.request.urlopen("https://api.datacite.org/dois/" + urllib.parse.quote(doi), timeout=30) as r:
        a = json.load(r)["data"]["attributes"]
    fam = lambda c: c.get("familyName") or c["name"].split(",")[0]
    return {"title": [a["titles"][0]["title"]], "author": [{"family": fam(c), "sequence": "first" if i == 0 else "additional"}
                                                             for i, c in enumerate(a["creators"])],
            "issued": {"date-parts": [[a["publicationYear"]]]}, "container-title": [a.get("publisher", "")]}


def first_family(authors_field):
    first = authors_field.split(" and ")[0].strip()
    fam = first.split(",")[0] if "," in first else first.split()[-1]
    return " ".join(norm(fam))


def check(key, e):
    doi = e.get("doi")
    if not doi:
        return "NO-DOI", "no DOI; check by hand"
    try:
        m = crossref(doi)
    except Exception as ex:  # network or 404
        return "FAIL", f"Crossref lookup failed: {ex}"
    problems = []
    title = (m.get("title") or [""])[0]
    if similar(e.get("title", ""), title) < 0.90:
        problems.append(f"title: bib '{e.get('title','')[:70]}' vs crossref '{title[:70]}'")
    year = None
    for k in ("published-print", "published-online", "issued", "created"):
        if m.get(k, {}).get("date-parts"):
            year = m[k]["date-parts"][0][0]
            break
    years = {str(m[k]["date-parts"][0][0]) for k in ("published-print", "published-online", "issued")
             if m.get(k, {}).get("date-parts") and m[k]["date-parts"][0][0]}
    if e.get("year") and e["year"] not in years:
        problems.append(f"year: bib {e['year']} vs crossref {sorted(years)}")
    authors = m.get("author") or []
    # Crossref sometimes lists a consortium (a "name" with no "family") ahead of the first person,
    # as for Grove et al. 2019. Compare against the first named person, and also accept a match with
    # the listed first author, so a consortium-first record neither hides nor fakes a mismatch.
    people = [a["family"] for a in authors if a.get("family")]
    listed_first = next((a.get("family") or a.get("name") for a in authors if a.get("sequence") == "first"),
                        (authors[0].get("family") or authors[0].get("name")) if authors else "")
    cr_first = people[0] if people else listed_first
    if e.get("author") and (cr_first or listed_first):
        bf = first_family(e["author"])
        cands = {" ".join(norm(x)) for x in (cr_first, listed_first) if x}
        if not any(bf == c or bf in c or c in bf for c in cands):
            problems.append(f"first author: bib '{bf}' vs crossref {sorted(cands)}")
    container = " ".join(m.get("container-title") or [])
    bib_container = e.get("journal") or e.get("booktitle") or ""
    if bib_container and container and similar(bib_container, container) < 0.6 \
            and not set(norm(bib_container)) <= set(norm(container)):
        problems.append(f"journal: bib '{bib_container}' vs crossref '{container}'")
    time.sleep(0.2)
    return ("FAIL", "; ".join(problems)) if problems else ("OK", f"{cr_first} {sorted(years)} {container[:40]}")


def main():
    bib = Path(sys.argv[1])
    entries = parse_bib(bib.read_text())
    failures = 0
    for key, e in entries.items():
        status, msg = check(key, e)
        failures += status == "FAIL"
        print(f"{status:6s} {key:28s} {msg}")
    if len(sys.argv) > 2:
        tex = Path(sys.argv[2]).read_text()
        cited = {k.strip() for grp in re.findall(r"\\cite[a-z]*\{([^}]*)\}", tex) for k in grp.split(",")}
        missing = cited - entries.keys()
        uncited = entries.keys() - cited
        for k in sorted(missing):
            print(f"FAIL   {k:28s} cited in tex but not in bib")
        for k in sorted(uncited):
            print(f"FAIL   {k:28s} in bib but never cited")
        failures += len(missing) + len(uncited)
    print(f"\n{len(entries)} entries, {failures} failures")
    sys.exit(1 if failures else 0)


if __name__ == "__main__":
    main()

File: src/verify_refs.R
## Verify every reference in a BibTeX file against Crossref, and every citation
## against the file. Mismatches fail loudly.
##
## Usage: Rscript src/verify_refs.R paper/references.bib [paper/itoju.tex]

suppressPackageStartupMessages({
  library(jsonlite)
  library(curl)
})

MAILTO <- "itoju-reference-check"

norm_words <- function(s) {
  s <- iconv(if (is.null(s)) "" else s, to = "ASCII//TRANSLIT")
  s <- gsub("\\\\[a-zA-Z]+|[{}\\\\$]", "", s)
  w <- strsplit(trimws(gsub("[^a-z0-9 ]+", " ", tolower(s))), "\\s+")[[1L]]
  w[nzchar(w)]
}

similar <- function(a, b) {
  x <- paste(norm_words(a), collapse = " ")
  y <- paste(norm_words(b), collapse = " ")
  if (!nzchar(x) && !nzchar(y)) return(1)
  1 - adist(x, y)[1, 1] / max(nchar(x), nchar(y), 1L)
}

parse_bib <- function(text) {
  entries <- list()
  ms <- gregexpr("(?s)@(\\w+)\\s*\\{\\s*([^,]+),(.*?)\\n\\}", text, perl = TRUE)[[1L]]
  if (ms[1L] == -1L) return(entries)
  lens <- attr(ms, "match.length")
  for (i in seq_along(ms)) {
    blk  <- substr(text, ms[i], ms[i] + lens[i] - 1L)
    kind <- tolower(sub("(?s)^@(\\w+).*$", "\\1", blk, perl = TRUE))
    key  <- trimws(sub("(?s)^@\\w+\\s*\\{\\s*([^,]+),.*$", "\\1", blk, perl = TRUE))
    fm   <- gregexpr("(\\w+)\\s*=\\s*(\\{(?:[^{}]|\\{(?:[^{}]|\\{[^{}]*\\})*\\})*\\}|\"[^\"]*\"|\\d+)",
                     blk, perl = TRUE)[[1L]]
    flen <- attr(fm, "match.length")
    fields <- list(type = kind)
    if (fm[1L] != -1L) for (j in seq_along(fm)) {
      piece <- substr(blk, fm[j], fm[j] + flen[j] - 1L)
      nm    <- tolower(sub("^(\\w+)\\s*=.*", "\\1", piece, perl = TRUE))
      val   <- trimws(sub("^\\w+\\s*=\\s*", "", piece))
      if (substr(val, 1L, 1L) %in% c("{", "\"")) val <- substr(val, 2L, nchar(val) - 1L)
      fields[[nm]] <- trimws(gsub("\\s+", " ", val))
    }
    entries[[key]] <- fields
  }
  entries
}

fetch_json <- function(url) {
  h <- new_handle(); handle_setheaders(h, "User-Agent" = sprintf("verify_refs/1.0 (%s)", MAILTO))
  r <- tryCatch(curl_fetch_memory(url, handle = h), error = function(e) NULL)
  if (is.null(r)) return(list(status = 0L, body = NULL))
  list(status = r$status_code,
       body = tryCatch(fromJSON(rawToChar(r$content), simplifyVector = FALSE), error = function(e) NULL))
}

crossref <- function(doi) {
  ## Crossref record, or a DataCite record mapped to the same fields (repository
  ## DOIs such as Stanford Digital Repository preprints are registered with
  ## DataCite, not Crossref).
  r <- fetch_json(paste0("https://api.crossref.org/works/", URLencode(doi, reserved = TRUE)))
  if (r$status == 200L && !is.null(r$body)) return(r$body$message)
  if (r$status != 404L && r$status != 0L)
    stop(sprintf("Crossref returned status %s for %s", r$status, doi), call. = FALSE)
  d <- fetch_json(paste0("https://api.datacite.org/dois/", URLencode(doi, reserved = TRUE)))
  if (d$status != 200L || is.null(d$body)) stop(sprintf("neither Crossref nor DataCite resolved %s", doi), call. = FALSE)
  a   <- d$body$data$attributes
  fam <- function(c) if (!is.null(c$familyName)) c$familyName else strsplit(c$name, ",")[[1L]][1L]
  list(title = list(a$titles[[1L]]$title),
       author = lapply(seq_along(a$creators), function(i)
         list(family = fam(a$creators[[i]]), sequence = if (i == 1L) "first" else "additional")),
       issued = list(`date-parts` = list(list(a$publicationYear))),
       `container-title` = list(if (is.null(a$publisher)) "" else a$publisher))
}

first_family <- function(authors_field) {
  first <- trimws(strsplit(authors_field, " and ", fixed = TRUE)[[1L]][1L])
  fam <- if (grepl(",", first, fixed = TRUE)) strsplit(first, ",")[[1L]][1L]
         else tail(strsplit(first, "\\s+")[[1L]], 1L)
  paste(norm_words(fam), collapse = " ")
}

check_entry <- function(key, e) {
  doi <- e$doi
  if (is.null(doi) || !nzchar(doi)) return(c("NO-DOI", "no DOI; check by hand"))
  m <- tryCatch(crossref(doi), error = function(ex) ex)
  if (inherits(m, "condition")) return(c("FAIL", sprintf("Crossref lookup failed: %s", conditionMessage(m))))
  problems <- character(0)
  title <- if (length(m$title)) m$title[[1L]] else ""
  if (similar(if (is.null(e$title)) "" else e$title, title) < 0.90)
    problems <- c(problems, sprintf("title: bib '%s' vs crossref '%s'",
                                    substr(if (is.null(e$title)) "" else e$title, 1, 70), substr(title, 1, 70)))
  years <- unique(unlist(lapply(c("published-print", "published-online", "issued"), function(k)
    if (!is.null(m[[k]]$`date-parts`)) as.character(m[[k]]$`date-parts`[[1L]][[1L]]))))
  if (!is.null(e$year) && nzchar(e$year) && !(e$year %in% years))
    problems <- c(problems, sprintf("year: bib %s vs crossref %s", e$year, paste(sort(years), collapse = ",")))
  authors <- if (is.null(m$author)) list() else m$author
  ## Crossref sometimes lists a consortium (a "name" with no "family") ahead of the
  ## first person, as for Grove et al. 2019. Compare against the first named person,
  ## and also accept a match with the listed first author, so a consortium-first
  ## record neither hides nor fakes a mismatch.
  people <- unlist(lapply(authors, function(a) a$family))
  listed_first <- NULL
  for (a in authors) if (identical(a$sequence, "first")) { listed_first <- if (!is.null(a$family)) a$family else a$name; break }
  if (is.null(listed_first) && length(authors))
    listed_first <- if (!is.null(authors[[1L]]$family)) authors[[1L]]$family else authors[[1L]]$name
  cr_first <- if (length(people)) people[1L] else listed_first
  if (!is.null(e$author) && nzchar(e$author) && (length(cr_first) || length(listed_first))) {
    bf    <- first_family(e$author)
    cands <- unique(vapply(Filter(Negate(is.null), list(cr_first, listed_first)),
                           function(x) paste(norm_words(x), collapse = " "), character(1L)))
    if (!any(vapply(cands, function(c) bf == c || grepl(bf, c, fixed = TRUE) || grepl(c, bf, fixed = TRUE), logical(1))))
      problems <- c(problems, sprintf("first author: bib '%s' vs crossref %s", bf, paste(sort(cands), collapse = ",")))
  }
  container     <- paste(unlist(m$`container-title`), collapse = " ")
  bib_container <- if (!is.null(e$journal)) e$journal else if (!is.null(e$booktitle)) e$booktitle else ""
  if (nzchar(bib_container) && nzchar(container) && similar(bib_container, container) < 0.6 &&
      !all(norm_words(bib_container) %in% norm_words(container)))
    problems <- c(problems, sprintf("journal: bib '%s' vs crossref '%s'", bib_container, container))
  Sys.sleep(0.2)
  if (length(problems)) c("FAIL", paste(problems, collapse = "; "))
  else c("OK", sprintf("%s %s %s", cr_first, paste(sort(years), collapse = ","), substr(container, 1, 40)))
}

main <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  if (!length(args)) stop("usage: Rscript src/verify_refs.R references.bib [paper.tex]", call. = FALSE)
  entries  <- parse_bib(paste(readLines(args[1L], warn = FALSE), collapse = "\n"))
  failures <- 0L
  for (key in names(entries)) {
    r <- check_entry(key, entries[[key]])
    failures <- failures + (r[1L] == "FAIL")
    cat(sprintf("%-6s %-28s %s\n", r[1L], key, r[2L]))
  }
  if (length(args) >= 2L) {
    tex    <- paste(readLines(args[2L], warn = FALSE), collapse = "\n")
    groups <- regmatches(tex, gregexpr("\\\\cite[a-z]*\\{[^}]*\\}", tex, perl = TRUE))[[1L]]
    cited  <- unique(trimws(unlist(strsplit(sub("^\\\\cite[a-z]*\\{", "", sub("\\}$", "", groups)), ","))))
    cited  <- cited[nzchar(cited)]
    missing <- sort(setdiff(cited, names(entries)))
    uncited <- sort(setdiff(names(entries), cited))
    for (k in missing) cat(sprintf("FAIL   %-28s cited in tex but not in bib\n", k))
    for (k in uncited) cat(sprintf("FAIL   %-28s in bib but never cited\n", k))
    failures <- failures + length(missing) + length(uncited)
  }
  cat(sprintf("\n%d entries, %d failures\n", length(entries), failures))
  quit(status = if (failures) 1L else 0L)
}

if (sys.nframe() == 0L) main()
